src/main/java/com/bobocode/blyznytsia/bibernate/persister/EntityPersisterImpl.java

```diff
@@ -0,0 +1,162 @@
+      fillSelectStatement(statement, value);
+      var rs = statement.executeQuery();
+      while (rs.next()) {
+        entities.add(resultSetMapper.mapToEntity(rs, entityType));
+      }
+      return entities;
+    });
+  }
+
+
+  @Override
+  public <T> Optional<T> findOneBy(Class<T> entityType, Field key, Object value) {
+    List<T> entitiesFound = findAll(entityType, key, value);
+    if (entitiesFound.size() > 1) {
+      throw new PersistenceException("Query returned more than one entity");
+    }
+    return entitiesFound.stream().findFirst();
+  }
+
+  @Override
+  public <T> Optional<T> findById(Class<T> entityType, Object id) {
+    if (id == null) {
+      throw new IllegalArgumentException("Id must not be null");
+    }
+    return findOneBy(entityType, EntityUtil.resolveEntityIdField(entityType), id);
+  }
+
+  @Override
+  public <T> T insert(T entity) {
+    Objects.requireNonNull(entity);
+    var insertStatementText = buildInsertStatement(entity.getClass());
+    return performWithinStatement(insertStatementText, true, statement -> {
+      fillInsertStatement(statement, entity);
+      statement.executeUpdate();
+      setGeneratedId(entity, statement.getGeneratedKeys());
+      return entity;
+    });
+  }
+
+  public void update(Object entity) {
+    Objects.requireNonNull(entity);
+    var updateStatementText = buildUpdateStatement(entity.getClass());
+    performWithinStatement(updateStatementText, statement -> {
+      fillUpdateStatement(statement, entity);
+      statement.executeUpdate();
+    });
+  }
+
+  @Override
+  public void delete(Object entity) {
+    Objects.requireNonNull(entity);
+    var deleteStatementText = buildDeleteStatement(entity.getClass());
+    performWithinStatement(deleteStatementText, statement -> {
+      fillDeleteStatement(statement, entity);
+      statement.executeUpdate();
+    });
+  }
+
+  private void performWithinStatement(String statement, StatementConsumer consumer) {
+    performWithinStatement(statement, false, consumer);
+  }
+
+  private <T> T performWithinStatement(String statementText, boolean generatedKeys, StatementFunction<T> function) {
+    log.info("Executing SQL statement: {}", statementText);
+    try (var statement = connection.prepareStatement(statementText,
+        generatedKeys ? Statement.RETURN_GENERATED_KEYS : Statement.NO_GENERATED_KEYS)
+    ) {
+      return function.apply(statement);
+    } catch (SQLException e) {
+      throw new PersistenceException("Failed performing SQL statement: " + statementText, e);
+    }
+  }
+
+  private void fillSelectStatement(PreparedStatement statement, Object value) throws SQLException {
+    fillStmtWildcards(statement, value);
+  }
+
+  private <T> void fillDeleteStatement(PreparedStatement statement, T entity) throws SQLException {
+    var id = getEntityIdValue(entity);
+    if (id == null) {
+      throw new PersistenceException("Cannot delete transient entity");
```

Comment:
Probably introduce TransientObjectException if it makes sense ...

Reply:
I think it does. Added `TransientEntityException`